### ISSUE TITLE
[java] UnsynchronizedStaticFormatter documentation and implementation wrong

### ISSUE THREAD

**Issue by boris-petrov** [pmd/pmd#1814]
I'm not exactly sure I'm correct but I wanted to note this.

In the [documentation](https://pmd.github.io/pmd-6.14.0/pmd_rules_java_multithreading.html#unsynchronizedstaticformatter) it is written that `If multiple threads must access a static formatter, the formatter must be synchronized either on method or block level.`. But mustn't the formatter be synchronized on the *class* instead of on the *instance* if it is a static one? Going through the Java source code of `SimpleDateFormat` for example I can see that this is indeed the case. If I am correct, than the example in the documentation (and the check itself) is wrong:

```java
public class Foo {
    private static final SimpleDateFormat sdf = new SimpleDateFormat();
    void bar() {
        sdf.format(); // poor, no thread-safety
    }
    synchronized void foo() {
        sdf.format(); // preferred
    }
}
```

Rather, it should be:

```java
public class Foo {
    private static final SimpleDateFormat sdf = new SimpleDateFormat();
    void bar() {
        sdf.format(); // poor, no thread-safety
    }
    void foo() {
        synchronized (Foo.class) {
            sdf.format(); // preferred
        }
    }
}
```

Am I mistaken somewhere?

**Comment by jsotuyod** [comment]
@boris-petrov you are absolutely right.

I'd not synchronize on the class nonetheless, as that may have unexpected side-effects, since outside classes / methods may synchronize on that too and cause contention (see Java Concurrency in Practice). I'd instead synchronize on `sdf` itself.

**Comment by boris-petrov** [comment]
@jsotuyod - yes, I agree, synchronizing on `sdf` is the better choice.

In that case, there is also a bug in the rule's implementation, not only the documentation - as it doesn't catch the example which is in the documentation - and it should. So perhaps remove the `in:documentation` label. :)

**Comment by jsotuyod** [comment]
@boris-petrov which example is not detected?

https://github.com/pmd/pmd/blob/601a6be306ae21499d8201f48d469be0cda6e78a/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/multithreading/xml/UnsynchronizedStaticFormatter.xml#L6-L19

is even a test case…

**Comment by boris-petrov** [comment]
@jsotuyod:

```java
public class Foo {
    private static final SimpleDateFormat sdf = new SimpleDateFormat();
    void bar() {
        sdf.format(); // poor, no thread-safety
    }
    synchronized void foo() {
        sdf.format(); // preferred
    }
}
```

This is the example in the documentation. It is wrong as we agreed, right? The detector doesn't catch it. The `sdf` variable should be synchronized on either `Foo.class` or `sdf` itself, *not* on `this` as is in the example. The detector should catch this.

**Comment by jsotuyod** [comment]
@boris-petrov ohh, right! I'm opening a separate issue for that one
